### DISCUSSION TITLE
sparse_set / storage iterators going from back to front

### DISCUSSION THREAD

**Discussion by KabelitzJ** [skypjack/entt#849] [Q&A] [answered]
Ive recently looked into the implementation off entt and dircovererd that the custom iterators for the sparse_set and the storage iterate from the back to the fron. What is the reasoning behind that?

**Comment by skypjack** [comment, marked as answer]
That's a trick to make it possible to add and remove elements during iterations.
For perf reasons, EnTT does a swap-and-pop when you delete a component. Imagine now to iterate from 0 to N. If you store aside the size, you can still _add_ new elements during iterations. However, if you remove an element, it gets super tricky and error prone.
What happens when you iterate backwards? Adding isn't a problem since the new elements go to the end while you move towards the head. Similarly, if you remove the current element, you swap-and-pop with the last one which is an already visited object. Also in this case, no errors nor risks.

**Reply by KabelitzJ** [reply]
That makes sense! Thank you very much for the indepth explanation
